rfcs/0049-flakes.md

```diff
@@ -0,0 +1,703 @@
+
+  ```
+  # nix build github:NixOS/patchelf
+  ```
+
+* For a reproducible result, you can also use a specific revision:
+
+  ```
+  # nix build nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06:hello
+  ```
+
+* To get information about a flake:
+
+  ```
+  # nix flake info nixpkgs
+  Description:   A collection of packages for the Nix package manager
+  Revision:      a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ...
+  ```
+
+* In addition to the global registry, there is a per-user
+  registry. This can be used to pin flakes to the current version:
+  ```
+  # nix flake pin nixpkgs
+  ```
+  or to a specific version:
+  ```
+  # nix flake add nixpkgs github:edolstra/nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ```
+
+* Flakes can have dependencies on other flakes. For example, [the
+  `patchelf`
+  flake](https://github.com/NixOS/patchelf/blob/master/flake.nix)
+  depends on the `nixpkgs` flake. To ensure reproducibility,
+  dependencies are pinned to specific versions using a *lock file*
+  (e.g. [for
+  patchelf](https://github.com/NixOS/patchelf/blob/master/flake.lock)). Lock
+  files are generated automatically.
+
+* Flakes are evaluated in pure mode, meaning they can't access
+  anything other than their own source or declared dependencies. This
+  allows `nix` to cache evaluation results.
+
+* The `nix` command is flake-based. For example,
+  ```
+  # nix build
+  ```
+  builds the flake in the current directory.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. It can be
+seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
+rather than a separate tool.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a Git repository that contains a file named `flake.nix` in
```

Comment:
Defining flakes explicitly as git repos seems like a useful stepping-stone toward a delightful future but I sense that it may even now be too restrictive for what I think even early-majority flake adopters will want, let alone what the long tail of users will want. 

As a result, would it be possible to sketch out at least a little bit, here or elsewhere, where the current design is driven to depend exclusively on git for short-term convenience, expediency, or concreteness vs more long-term reasons so that we can think together about how the needs or aspirations those reasons represent might be addressed in non-git scenarios e.g., involving different network, security, compatibility, usability, or scale considerations?

Reply:
To add a couple of high-level use cases motivating the question beyond more abstract design value judgments about configurability and coupling:

* while acknowledging the importance of human-meaningful flake metadata for good UX, as a user experimenting with flakes for the first time, I would like to be able to define flakes in local directories on my filesystem that are not yet version-controlled or that are version-controlled with something other than git

* as a user who occasionally interacts with monorepos, some of them large, some of them not based on git, or some of them based on git but using it in non-conventional ways, I would like flakes to be less tightly coupled not only with git but also with the idea that relevant units of code live *precisely* in "git repositories made available by a network service via URLs"

* finally, as a user interested in using nix flakes in diverse environments including on hardware or operating systems where git may not be available, as well as with colleagues who are not yet familiar with git or where software is being managed by other means, I would like the flakes design to be a little bit more agnostic about how flakes and their supporting resources are organized, aggregated, and transported even if awesome support for git and GitHub remain top-priority concerns.

Reply:
There is nothing about flakes that makes them inherently tied to Git or GitHub. It's just that those are the only two mechanisms implemented at the moment. Mercurial and tarball support would be straightforward to add.